File: sources/base/ui/src/main/java/com/egoriku/ui/arch/pvm/BaseViewModel.java
package com.egoriku.ui.arch.pvm;

import android.arch.lifecycle.ViewModel;

public final class BaseViewModel<V extends BaseContract.View, P extends BaseContract.Presenter<V>>
        extends ViewModel {

    private P presenter;

    public void setPresenter(P presenter) {
        if (this.presenter == null) {
            this.presenter = presenter;
        }
    }

    public P getPresenter() {
        return presenter;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        presenter.onPresenterDestroy();
        presenter = null;
    }
}

File: sources/base/ui/src/main/java/com/egoriku/ui/arch/pvm/BaseViewModel.kt
package com.egoriku.ui.arch.pvm

import android.arch.lifecycle.ViewModel

class BaseViewModel<V : BaseContract.View, P : BaseContract.Presenter<V>> : ViewModel() {

    var presenter: P? = null
        set(value) {
            if (presenter == null) {
                field = value
            }
        }

    override fun onCleared() {
        super.onCleared()
        presenter!!.onPresenterDestroy()
        presenter = null
    }
}